railties/lib/rails/generators/rails/app/templates/config/puma.rb.tt

```diff
@@ -2,38 +2,53 @@
-# and maximum; this matches the default thread size of Active Record.
-max_threads_count = ENV.fetch("RAILS_MAX_THREADS") { 5 }
-min_threads_count = ENV.fetch("RAILS_MIN_THREADS") { max_threads_count }
-threads min_threads_count, max_threads_count
-
-# Specifies that the worker count should equal the number of processors in production.
-if ENV["RAILS_ENV"] == "production"
+rails_env = ENV.fetch("RAILS_ENV", "development")
+
+# Puma starts a configurable number of processes (workers) and each process
+# serves each request in a thread from an internal thread pool.
+#
+# As a rule of thumb, increasing the number of threads will increase how much
+# traffic a given process can handle (throughput), but due to Ruby's
+# Global VM Lock (GVL) it has diminishing returns, and will degrade the
+# response time (latency) of the application.
+#
+# The ideal number of threads per worker depend both how much time the
+# application spend waiting for IO operations, and how much you wish to
+# to prioritize throughput over latency.
+#
+# The default is set to 3 threads as it's deemed a decent compromise between
+# throughput and latency for the average Rails application.
+#
+# Any libraries that use a connection pool or another resource pool should
+# be configured to provide at least as many connections as the number of
+# threads. This includes Active Record's `pool` parameter in `database.yml`.
+threads_count = ENV.fetch("RAILS_MAX_THREADS") { 3 }
+threads threads_count, threads_count
+
+if rails_env == "production"
+  # Specifies that the worker count should equal the number of processors in production.
+  #
+  # The number of processor is detected using various platform dependent methods, but
+  # note that this isn't very reliable, especially in containerized environments like Docker.
+  # As such, it is recommended to set it manually.
   require "concurrent-ruby"
   worker_count = Integer(ENV.fetch("WEB_CONCURRENCY") { Concurrent.physical_processor_count })
   workers worker_count if worker_count > 1
 end
 
-# Specifies the `worker_timeout` threshold that Puma will use to wait before
-# terminating a worker in development environments.
-worker_timeout 3600 if ENV.fetch("RAILS_ENV", "development") == "development"
-
 # Specifies the `port` that Puma will listen on to receive requests; default is 3000.
 port ENV.fetch("PORT") { 3000 }
 
 # Specifies the `environment` that Puma will run in.
-environment ENV.fetch("RAILS_ENV") { "development" }
-
-# Specifies the `pidfile` that Puma will use.
-if ENV["PIDFILE"]
-  pidfile ENV["PIDFILE"]
-else
-  pidfile "tmp/pids/server.pid" if ENV.fetch("RAILS_ENV", "development") == "development"
-end
+environment rails_env
 
 # Allow puma to be restarted by `bin/rails restart` command.
 plugin :tmp_restart
+
+if rails_env == "development"
+  # Specifies the `worker_timeout` threshold that Puma will use to wait before
+  # terminating a worker in development environments.
```

Comment:
Nitpick, but should we clarify that this is only if the worker [hangs or is dead](https://github.com/puma/puma/blob/99f83c50fbb712b0987667f3533cce4ea925b2da/lib/puma/dsl.rb#L908)? Just so there's no misconception that it's an idle timeout.

Better yet would be to clarify that this is configured to make debugging easier (using `pry` and friends). I'm guessing most users opt for single mode in dev though...

Reply:
Hum, good catch. Based on the comment I think the intent was to use `worker_shutdown_timeout`, let me git blame to see.

Reply:
Nevermind, it's to avoid getting the worker killed if the process is suspended by a debugger, hence is no longer regularly pining the cluster process: https://github.com/rails/rails/commit/b43ede6b9d4dd88300dbe541a61108d86ff11a32

I'll see if I can clarify this comment.

Reply:
Updated. Let me know if it's still not clear.

Reply:
Short and sweet. LGTM.